feat(unity): Added Programmatic Configuration (#4742)
getsentry/sentry-docs: src/platforms/unity/usage/programmatic-options.mdx

---
title: Programmatic Configuration
description: "Learn how to modify Sentry Unity options programmatically."
sidebar_order: 100
---

> The Unity SDK contains multiple other SDKs under the hood, to provide native support. By default, we initialize those before the Unity game itself starts to collect errors that occur during startup and would get missed otherwise. The setup for those SDKs happens during build time which is the reason why we rely on the options being set in the editor configuration window and saved to `Assets/Resources/Sentry/SentryOptions.asset`. To provide a way to modify options programmatically we've added `ScriptableOptionsConfiguration` to the `Option Config` tab. Note that all changes to the options here are made in the C# layer during runtime, after the native SDKs already initialized and will not affect native crashes. The native SDKs only take options defined in the editor configuration window.

Reviewer: @bitsandfoxes accepted the change we did together on the call. But as discussed further these would need to be moved so won't merge this for now

```csharp
using Sentry.Unity;
using UnityEngine;

[CreateAssetMenu(fileName = "Assets/Resources/Sentry/SentryOptionsConfiguration", menuName = "Sentry/SentryOptionsConfiguration", order = 999)]
public class SentryOptionsConfiguration : ScriptableOptionsConfiguration
{
    public override void Configure(SentryUnityOptions options)
    {
    }
}
```

> The method `Configure` gets called right before the .NET SDK gets initialized and provides you with a place to i.e. implement your own filtering of events via the [BeforeSend](/platforms/unity/configuration/filtering/#using-platformidentifier-namebefore-send-) callback.

Reviewer 2: I'm not sure what you're using "i.e." here to mean. I've taken a guess.

```suggestion
The method `Configure` is called right before the .NET SDK is initialized and provides you with a place to, for example, implement your own filtering of events using the [BeforeSend](/platforms/unity/configuration/filtering/#using-platformidentifier-namebefore-send-) callback.
```

---
title: Programmatic Configuration
description: "Learn how to modify Sentry Unity options programmatically."
sidebar_order: 100
---

> In order to provide native crash support, the Sentry SDK for Unity include platform-specific (aka Native) SDKs, such as the [Android](/platforms/android/), [Apple](/platforms/apple/guides/ios/) and [Native](/platforms/native/). 

Reviewer 2: ```suggestion
In order to provide native crash support, the Sentry SDK for Unity includes platform-specific (that is, Native) SDKs, such as the [Android](/platforms/android/), [Apple](/platforms/apple/guides/ios/), and [Native](/platforms/native/). 
```

> By default, we initialize the native SDKs before Unity itself runs. This allows us to capture bugs/crashes of the engine itself, any native plugin and also anything written in C#. The setup for the native SDKs happens during build time, which is the reason we rely on the options being set in the Sentry editor configuration window and saved to `Assets/Resources/Sentry/SentryOptions.asset`. To provide a way to modify options programmatically, we've added `ScriptableOptionsConfiguration` to the `Option Config`.

Reviewer 2: ```suggestion
By default, we initialize the native SDKs before Unity itself runs. This allows us to capture bugs/crashes of the engine itself, any native plugin, and anything written in C#. The setup for the native SDKs happens during build time, which is why we rely on the options being set in the Sentry editor configuration window and saved to `Assets/Resources/Sentry/SentryOptions.asset`. To provide a way to modify options programmatically, we've added `ScriptableOptionsConfiguration` to the `Option Config`.
```

> Note that all changes to the options object done through `ScriptableOptionsConfiguration` will not affect events coming from the native layer. That also means that a `BeforeSend` callback will only modify events coming from C# scripts. The native SDKs only take options defined in the Sentry editor configuration window.

Reviewer 2: ```suggestion
Note that changes to the options object done through `ScriptableOptionsConfiguration` do not affect events coming from the native layer. That also means that a `BeforeSend` callback will only modify events coming from C# scripts. The native SDKs only take options defined in the Sentry editor configuration window.
```
